### PR TITLE
Avoid zero-length memcpy in formatting

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This has two separate and somewhat orthogonal commits. The first change adjusts the ToString general impl for all types that implement Display; it no longer uses the full format machinery, rather directly falling onto a `std::fmt::Display::fmt` call. The second change directly adjusts the general core::fmt::write function which handles the production of format_args! to avoid zero-length push_str calls.

Both changes target the fact that push_str will still call memmove internally (or a similar function), as it doesn't know the length of the passed string. For zero-length strings in particular, this is quite expensive, and even for very short (several bytes long) strings, this is also expensive. Future work in this area may wish to have us fallback to write_char or similar, which may be cheaper on the (typically) short strings between the interpolated pieces in format_args!.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
@bors try @rust-timer queue

It may be advantageous to consider guarding the write_str calls in the formatting machinery here with a nonzero length check too, or trying to rework formatting machinery to avoid empty pieces entirely, but those are less clear wins I suspect.

File: library/core/src/fmt/mod.rs [hunk 1]
```diff
@@ -221,6 +221,19 @@ pub struct Formatter<'a> {
...
 }
 
+/// An impl.
+#[stable(feature = "rust1", since = "1.0.0")]
+pub fn new_fmt_buf_default<'a>(buf: &'a mut (dyn Write + 'a)) -> Formatter<'a> {
+    Formatter {
+        flags: 0,
+        width: None,
+        precision: None,
+        buf: buf,
```
Comment (joshtriplett) [inline library/core/src/fmt/mod.rs:231]:
```suggestion
        buf,
```

**Comment by joshtriplett** [comment]
Nice performance win!

It seems unfortunate to have to have a public function just because it lives in core and needs to be called from alloc. Is there any way we could avoid exposing that function?

**Comment by Mark-Simulacrum** [comment]
I don't think exposing the function is necessarily avoidable, but it is worth noting there have been several requests to create/store `Formatter`s; we may wish to expose that functionality publicly eventually - https://github.com/rust-lang/rust/issues/74870, https://github.com/rust-lang/rust/issues/46591. For now I've moved it to a method and marked it as unstable and doc(hidden) -- so we're barely exposing it.

@bors try @rust-timer queue -- added an is_empty guard on the push_str calls, we'll see if that helps or hurts here. Maybe that would be sufficient and we don't need the extra function at all, though I'd guess to_string() is a bit more efficient without going through the format machinery at all.

**Comment by Mark-Simulacrum** [comment]
Alright, that seemed to just make this more of an improvement. I'll try an intermediate version (i.e., just the second commit) later too, and then we can decide which pieces to land.

**Comment by Mark-Simulacrum** [comment]
Seems like both optimizations are complementary, so I think we should do both. Combined perf results are [here](https://perf.rust-lang.org/compare.html?start=44ec846f4ea68ffa6d06e7d68f078bd3cc59d4ec&end=1dcdf9bc01bf15dfd254ba6643d4747b20b4cb16), the first commit is evaluated individually [here](https://perf.rust-lang.org/compare.html?start=fe72845f7bb6a77b9e671e6a4f32fe714962cec4&end=021776915c55787521fe6b579f88425c7d9bff19) and the second one [here](https://perf.rust-lang.org/compare.html?start=9964284fed181676300aad693449f5b751e35ff2&end=3caeb07804c31d2ad5de523df7e6348f65c4cf07).

**Comment by scottmcm** [comment]
This change seems reasonable, given the perf results, so
@bors r+

Does make me wonder what it is about the `*-doc` builds that they're hitting this so consistently, though...

**Comment by Mark-Simulacrum** [comment]
-doc builds are just generating documentation which is naturely more dominated by string formatting; this is likely to improve the performance of any string formatting which has a zero-length string in between or before the first interpolated (e.g., `{}`) piece.

**Comment by scottmcm** [comment]
MacOS builder lost networking.

@bors retry

**Pull request merged**